### ISSUE TITLE
GPU solver on multi-GPU system memory allocation issue

### ISSUE THREAD

**Issue by cjosey** [mit-crpg/OpenMOC#192]
So, I was testing #162 on the GPU.  The only machine with GPUs I have available has 3 GPU installed simultaneously:
ID 0 is a GTX Titan X
ID 1 is a Quadro K600
ID 2 is a Tesla K20c
I compiled the current develop branch of OpenMOC with CUDA 7.5, and ran c5g7-gpu.py.  I get the following result:
```
[  NORMAL ]  Computing the eigenvalue...
Traceback (most recent call last):
  File "c5g7-gpu.py", line 328, in <module>
    solver.computeEigenvalue(max_iters)
  File "/home/cjosey/.local/lib/python3.4/site-packages/openmoc/cuda/openmoc_cuda.py", line 153, in computeEigenvalue
    def computeEigenvalue(self, *args, **kwargs) -> "void" : return _openmoc_cuda.Solver_computeEigenvalue(self, *args, **kwargs)
RuntimeError: Could not allocate memory for fluxes on GPU
```
I then traced that error message to ``initializeFluxArrays``, specifically, the instant ``_boundary_flux.resize(size);`` is called, an error is thrown and the code fails.

As a slight modification, I modified the code so that ``cudaSetDevice(n);`` was called at the beginning of ``initializeFluxArrays``.  If ``n`` is set to 0 or 2, it fails identically to the above.  If it's set to 1, I get the following (with some debug code to output the size of the arrays allocated and GPU used):

```
[  NORMAL ]  Computing the eigenvalue...
CUDA Device:             Quadro K600
76440
936012
[  NORMAL ]  Initializing FSRs on the GPU...
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  cudaFree in free: an illegal memory access was encountered
Aborted (core dumped)
```

I am unfamiliar with GPU coding, so I do not know if this is my fault, the code's fault, or some combination of the two.  @wbinventor, any ideas?

**Comment by wbinventor** [comment]
First, the Python interface in "openmoc.cuda" already includes an
"attach_gpu(...)" routine to select which GPU you wish to use.

I tested this code out with a GPU using CUDA 7.0 an AWS server last night
and it worked fine. I think your comments here should be raised as a
separate issue from this PR. I'm not sure how to fix them, but have seen
issues similar to them in the past when debugging CUDA code. You might try
running the "c5g7-gpu.py" case to see if it works or not.

**Comment by cjosey** [comment]
That is what I tried, as I said above.  I was only testing whether or not I could get GPU working on this machine.  

I ran ``CUDA_VISIBLE_DEVICES="0" python3 c5g7-gpu.py`` as well, and it didn't fix it.  As such, it may not be multiple GPU but CUDA 7.5 or something else.

**Comment by wbinventor** [comment]
I don't believe anyone has tried CUDA 7.5, so that may be it.

On Tue, Aug 4, 2015 at 10:32 AM, Colin Josey <notifications@github.com>
wrote:

> That is what I tried, as I said above. I was only testing whether or not I
> could get GPU working on this machine.
>
> I ran CUDA_VISIBLE_DEVICES="0" python3 c5g7-gpu.py as well, and it didn't
> fix it. As such, it may not be multiple GPU but CUDA 7.5 or something else.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/issues/192#issuecomment-127686280>.
>



-- 
Will Boyd
Nuclear Science & Engineering
Massachusetts Institute of Technology
Cell: 423.413.8469
wbinventor@gmail.com

**Comment by cjosey** [comment]
I was able to get it to run, but only on the K600, and not the newer hardware by adding ``attach_gpu(1)`` to the code.  Further, I tested with CUDA 7.0, and the issue is identical.

I did fix it though.  I changed the compiler flags such that instead of:
```
-gencode=arch=compute_20,code=sm_20
-gencode=arch=compute_30,code=sm_30
```
I used
``--gpu-architecture=compute_20 --gpu-code=compute_20``

This flag replaces the GPU code with JIT compiled code with a minimum support of compute_20.  Running this, it works just fine on the Titan X.  @jtramm said he had a "good" set of compiler flags he recommends, so I'll do some testing on this front.
